pkg/cloudprovider/providers/gce/kms.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce
+
+import (
+	"encoding/base64"
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/cloudprovider"
+
+	cloudkms "google.golang.org/api/cloudkms/v1"
+	"google.golang.org/api/googleapi"
+)
+
+const (
+	// KMSServiceName is the name of the cloudkms provider registered by this cloud.
+	KMSServiceName = "gcp-cloudkms"
+
+	defaultGKMSKeyRing = "google-container-engine"
+)
+
+// gkmsConfig contains the GCE specific KMS configuration for setting up a KMS connection.
+type gkmsConfig struct {
+	// location is the KMS location of the KeyRing to be used for encryption. The default value is "global".
+	// It can be found by checking the available KeyRings in the IAM UI.
+	// This is not the same as the GCP location of the project.
+	// +optional
+	Location string
+	// keyRing is the keyRing of the hosted key to be used. The default value is "google-kubernetes".
+	// +optional
+	KeyRing string
+	// cryptoKey is the name of the key to be used for encryption of Data-Encryption-Keys.
+	CryptoKey string
+}
+
+// gkmsService provides Encrypt and Decrypt methods which allow cryptographic operations
+// using Google Cloud KMS service.
+type gkmsService struct {
+	parentName      string
+	cloudkmsService *cloudkms.Service
+}
+
+// KeyManagementService provides a named key management service supported by the cloud.
+func (gce *GCECloud) KeyManagementService(name string) (cloudprovider.KeyManagementService, error) {
+	return nil, fmt.Errorf("cloud %q does not support KeyManagementService %q", gce.ProviderName(), name)
```

Comment:
why do you add this to the interface if you don't implement it for GCE? Is this expected?

Reply:
The last commit (WIP) removed the need for this method in the interface. But that commit adds a new dependency to cloudproviders, so I'm still waiting on more discussion on that before I delete this code. [PTAL](https://github.com/kubernetes/kubernetes/pull/48574#issuecomment-321847261).

I have thought up of a way to avoid that dependency, was planning to write the current design in a design doc with a diagram of dependencies and requirements of plugins. Would get that done over the weekend hopefully.